src/base.cc

```diff
@@ -1,6 +1,61 @@
 #include "astshim/base.h"
 
+#include <sstream>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
 namespace ast {
+namespace {
+
+    static std::ostringstream errorMsgStream;
+
+    /*
+    Write an error message to `errorMsgStream`
+
+    Intended to be registered as an error handler to AST by calling `astSetPutErr(reportError)`.
+    */
+    void reportError(int errNum, const char * errMsg) {
+        errorMsgStream << errMsg;
+    }
+
+    /*
+    Instantiate this class to register `reportError` as an AST error handler.
+    */
+    class ErrorHandler {
+    public:
+        ErrorHandler() {
+            astSetPutErr(reportError);
+        }
+
+        std::string getErrMsg() {
+            std::string errMsg = errorMsgStream.str();
+            errorMsgStream.clear();
+            if (errMsg.empty()) {
+                errMsg = "Failed with AST status = " + std::to_string(astStatus);
+            }
+            astClearStatus;
+            return errMsg;
+        }
+    };
+
+}  // <anonymous>
+
+void assertOK(AstObject * rawPtr1, AstObject * rawPtr2) {
+    // Construct ErrorHandler once, the first time this function is called.
+    // This is done to initialize `errorMsgStream` and register `reportError` as the AST error handler.
+    // See https://isocpp.org/wiki/faq/ctors#static-init-order
+    static ErrorHandler * errHandler = new ErrorHandler();
```

Comment:
@r-owen We don't have a formal system for that, but it's an interesting concept. Probably combining ls.st with an auto-archive service that fails over if the original link 404s. I'll put that in my idea bucket. Something probably exists off-the-shelf for this. In the meantime, webarchive.org can save us if the link goes stale.